### PR TITLE
feat(prefer-implicit-assert): adding new rule

### LINKED ISSUE #743: Unnecessary presence checks on findBy queries
### Name for new rule

no-find-by-presence

### Description of the new rule

The rule should avoid things like expect(await screen.findBy*()).toBeInTheDocument(), because that's functionally the same as await screen.findBy*().

Perhaps this is a new rule (I proposed a name) or maybe it should just be part of the existing find-by rule.

### Testing Library feature

await findBy

### Testing Library framework(s)

All, presumaly.

### What category of rule is this?

Suggests an alternate way of doing something

### Optional: other category of rule

_No response_

### Code examples

expect(await screen.findByRole("button")).toBeInTheDocument();
expect(await screen.findByText("hello")).toBeDefined();
expect(await screen.findByLabel("zoop!")).toBeTruthy();

### Anything else?

I am happy to contribute this rule - but if so, it may take me a month or two to do so, as things are busy. Please let me know if you'd like me to contribute or not.

### Do you want to submit a pull request to make the new rule?

Yes

### PR DESCRIPTION
**Pull request by adevick**
## Checks

- [x] I have read the [contributing guidelines](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/CONTRIBUTING.md).

## Changes

- Adding a new rule to prefer implicit assertions when using `findBy*` and `getBy*` queries

## Context

Closes #743

### CODE REVIEW

**Comment by adevick** [comment]
@Belco90 just checking in thanks!

**Comment by Belco90** [comment]
I'll try to review this next week.

File: lib/rules/prefer-implicit-assert.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+			},
+			'Program:exit'() {
+				let isAsyncQuery = true;
+				findQueryCalls.forEach((queryCall) => {
+					const node: TSESTree.Identifier | TSESTree.Node | undefined =
+						isCalledUsingSomeObject(queryCall) ? queryCall.parent : queryCall;
+
+					if (node) {
+						if (isCalledInExpect(node, isAsyncQuery)) {
+							if (usesPresenceAssertion(node, isAsyncQuery))
```
Comment (Belco90) [inline lib/rules/prefer-implicit-assert.ts:153]: The `helpers` arg already provides a `isPresenceAssert` util. Can you use that one instead?

File: lib/rules/prefer-implicit-assert.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+								return reportError(context, node, 'findBy*');
+							if (usesNotPresenceAssertion(node, isAsyncQuery))
```
Comment (Belco90) [inline lib/rules/prefer-implicit-assert.ts:155]: The `helpers` arg already provides a `isAbsenceAssert` util. Can you use that one instead?

File: lib/rules/prefer-implicit-assert.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+				}
+				if (helpers.isGetQueryVariant(node)) {
+					getQueryCalls.push(node);
+				}
+			},
+			'Program:exit'() {
+				let isAsyncQuery = true;
+				findQueryCalls.forEach((queryCall) => {
+					const node: TSESTree.Identifier | TSESTree.Node | undefined =
+						isCalledUsingSomeObject(queryCall) ? queryCall.parent : queryCall;
```
Comment (Belco90) [inline lib/rules/prefer-implicit-assert.ts:149]: You have the `getReferenceNode` util in the node-utils folder to do this in a generic way. Can you use that one instead?

**Review by Belco90 — CHANGES_REQUESTED** [review]
Looks good in general! I requested a couple of changes that should be simple to apply, and should be ready to go.

File: docs/rules/prefer-implicit-assert.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+expect(queryByText('foo')).not.toBeInTheDocument();
+```
+
+## When Not To Use It
+
+If you prefer to use `getBy*` & `findBy*` queries with explicitly asserting existance of elements, then this rule is not recommended. Instead check out this rule [prefer-explicit-assert](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/prefer-explicit-assert.md)
+
+## Further Reading
+
+- [getBy query](https://testing-library.com/docs/dom-testing-library/api-queries#getby)
```
[the remark this answers is not in the document]
Comment (adevick) [reply docs/rules/prefer-implicit-assert.md:56]: added.

File: tests/lib/rules/prefer-implicit-assert.test.ts [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+			code: `const something = query${queryMethod}('qux');`,
+		})),
+	],
+	invalid: [
+		...COMBINED_QUERIES_METHODS.map(
+			(queryMethod) =>
+				({
+					code: `expect(await find${queryMethod}('qux')).toBeInTheDocument();`,
+					errors: [
+						{
```
[the remark this answers is not in the document]
Comment (adevick) [reply tests/lib/rules/prefer-implicit-assert.test.ts:56]: added

File: lib/rules/prefer-implicit-assert.ts (hunk 1 above)
Comment (adevick) [reply lib/rules/prefer-implicit-assert.ts:153]: done

File: lib/rules/prefer-implicit-assert.ts (hunk 2 above)
Comment (adevick) [reply lib/rules/prefer-implicit-assert.ts:155]: done

File: lib/rules/prefer-implicit-assert.ts (hunk 3 above)
Comment (adevick) [reply lib/rules/prefer-implicit-assert.ts:149]: swapped it out but with the `||  isCallExpression` [here](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/lib/node-utils/index.ts#L405C1-L406C1) my tests were failing.
Thoughts?

**Comment by adevick** [comment]
@Belco90 Hey thanks for the review, fixed all but the last comment.

File: lib/rules/prefer-implicit-assert.ts (hunk 3 above)
Comment (Belco90) [reply lib/rules/prefer-implicit-assert.ts:100]: Mmm, I see. I can't see right now why that can cause your tests to fail, so let's leave it as it is.

**Review by Belco90 — APPROVED** [review]
LGTM, but it seems some extra tests to reach 90% of coverage.

**Review by Belco90 — APPROVED** [review]

**Comment by Belco90** [comment]
@all-contributors please add @adevick for code, test, docs

**Pull request merged**